interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,273 @@
+  // TODO: make ports arguments (and probably loopback/ip as well)
+  private String loopback = "127.0.0.1";
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int queueLength;
+  private int chunkSize;
+  private int bandwidth;
+  private long latency;
+  private volatile boolean shutDown;
+  private ServerSocket clientAcceptor;
+  private Socket serverSock;
+  private Socket clientSock;
+  private List<Thread> workers = new ArrayList<Thread>();
+
+  /**
+   * Returns a new TrafficControlProxy with default bandwidth and latency.
+   */
+  public TrafficControlProxy() {
+    this(DEFAULT_BAND_BPS, DEFAULT_DELAY_NANOS, TimeUnit.NANOSECONDS);
+  }
+
+  /**
+   * Returns a new TrafficControlProxy with bandwidth set to targetBPS, and latency set to
+   * targetLatency in latencyUnits.
+   */
+  public TrafficControlProxy(int targetBps, int targetLatency, TimeUnit latencyUnits) {
+    checkArgument(targetBps > 0);
+    checkArgument(targetLatency > 0);
+    bandwidth = targetBps;
+    // divide by 2 because latency is applied in both directions
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+  }
+
+  /**
+   * Convenience method to change bandwidth/latency without restarting.
+   */
+  public void reset(int targetBps, int targetLatency, TimeUnit latencyUnits) throws IOException {
+    checkArgument(targetBps > 0);
+    checkArgument(targetLatency > 0);
+    shutDown();
+    bandwidth = targetBps;
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+    workers.clear();
+    shutDown = false;
+    start();
+  }
+
+  /**
+   * Waits for client and server and start reader/writer threads.
+   */
+  public void start() throws IOException {
+    // ClientAcceptor uses a ServerSocket server so that the client can connect to the proxy as it
+    // normally would a server. serverSock then connects the server using a regular Socket as a
+    // client normally would.
+    logger.info("Starting new proxy on port " + proxyPort + " with Queue Length " + queueLength);
+    clientAcceptor = new ServerSocket();
+    clientAcceptor.setReuseAddress(true);
+    clientAcceptor.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = clientAcceptor.accept();
+    serverSock = new Socket();
+    serverSock.connect(new InetSocketAddress(loopback, serverPort));
+
+    startWorkers();
+  }
+
+  /** Interrupt all workers and close sockets. */
+  public void shutDown() throws IOException {
+    // TODO: Handle case where a socket fails to close, therefore blocking the others from closing
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    clientAcceptor.close();
+    clientSock.close();
+    serverSock.close();
+    for (Thread worker : workers) {
```

Comment:
Do another loop over the workers and join() on them? It seems like `reset()` requires all the threads to be exited by the time this method returns.